### PR TITLE
Handle encrypted files in "import key" file selection

### PR DESCRIPTION
**Pull request by runnerway**
This is a solution for:
https://github.com/open-keychain/open-keychain/issues/1673

### CODE REVIEW

**Comment by runnerway** [comment]
The only thing left is that some activity need to close themselves after the operation is completed.
I will address this probably tomorrow. Then it will be ready.

**Comment by runnerway** [comment]
Now it works pretty well.
The user experience is good and all seems in place.

**Comment by Valodim** [comment]
Awesome lookin' PRs. They also aren't trivial to review, please excuse the delay. I'll try to get to them tomorrow :+1:

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java [hunk 1]
```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
...
+            if (br.read(buffer, 0, length) == length) {
+                isEncrypted = new String(buffer).equals(header);
+            }
+        } catch (IOException e) {
+            throw e;
+        } finally {
+            try {
+                br.close();
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: Log.e

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java [hunk 2]
```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
...
+            br = new BufferedReader(new InputStreamReader(is));
+
+            String header = "-----BEGIN PGP MESSAGE-----";
+            int length = header.length();
+            char[] buffer = new char[length];
+            if (br.read(buffer, 0, length) == length) {
+                isEncrypted = new String(buffer).equals(header);
+            }
+        } catch (IOException e) {
+            throw e;
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: isn't this a nop? missing a Log.e mayhaps?

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java [hunk 3]
```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
...
+            isEncrypted = isEncryptedFile(mCurrentUri);
+        } catch (IOException e) {
+            e.printStackTrace();
+            Notify.create(mImportActivity, R.string.error_bad_data, Style.ERROR).show();
+            return;
+        }
+
+        if (isEncrypted) {
+            Intent intent = new Intent(getActivity(), DecryptActivity.class);
+            intent.setAction(DecryptActivity.ACTION_DECRYPT_FILE_AND_RETURN);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: I think this should be handled directly as part of the decryption operation, no need to roundtrip. That way we get the same behavior with encrypted keys regardless of what activity the input comes from and the activity stack. Just `startActivity`, `finish` and proceed from there.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java (hunk 1 above)
Comment (runnerway) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: Yes. Will fix it.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java (hunk 2 above)
Comment (runnerway) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: Yes. Will fix it.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java (hunk 3 above)
Comment (runnerway) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: Yes. That was my first try, but i spent a lot of time on it with not very good results regarding to user experience.
I thought the roundtrip was better for some reasons.

But i've just had some new ideas, i will try them and report back.
Thank you for the review.

**Comment by runnerway** [comment]
I've fixed the little problems you pointed out and i tried again to do without roundtrip.
Updated PR to new version. (I still have the older PR on my pc if we want to go back to that)

We have mainly 2 problems:
1) We can't "finish" the activity because otherwise we lose the permission to access the Uri
(http://developer.android.com/guide/topics/providers/content-provider-basics.html#Intents)
2) We need to close all the activities opened during the process to import encrypted keys, as soon as the operation is finished.

So i propose to do as in the latest version of this PR.
After we finish importing new keys, go back to MainActivity. (If not specified otherwise)
This would solve problem n.2 and it makes sense to me. 
Probably the user will go back to MainActivity to see the just imported keys anyway. 

I'm a bit tired now. I hope i didn't forget anything important.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java [hunk 4]
```diff
@@ -419,6 +419,16 @@ protected void handleResult(ImportKeyResult result) {
             return;
         }
 
+        if (result.isOkNew() || result.isOkUpdated()) {
+            Preferences prefs = Preferences.getPreferences(this);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java]: extract into method please

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java [hunk 5]
```diff
@@ -419,6 +419,16 @@ protected void handleResult(ImportKeyResult result) {
...
+            prefs.setFirstTime(false);
+
+            Intent intent = new Intent(this, MainActivity.class);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+            startActivity(intent);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java]: should this `finish()`?

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java [hunk 6]
```diff
@@ -121,4 +139,102 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+            Intent intent = new Intent(mImportActivity, DecryptActivity.class);
+            intent.setAction(Intent.ACTION_VIEW);
+            intent.setData(mCurrentUri);
+            startActivity(intent);
+        } else {
+            mImportActivity.loadCallback(new BytesLoaderState(null, mCurrentUri));
+        }
+    }
+
+    private boolean isEncryptedFile(Uri uri) throws IOException {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java]: make static and move to `FileHelper` please

**Comment by Valodim** [comment]
Good point about the uri permissions. Just tested it, look and feel seems good to me :+1:

**Comment by runnerway** [comment]
I will submit an updated PR with your notes fixed later today.
Thank you for the review.

**Comment by runnerway** [comment]
This should be ready for review. :)

**Comment by dschuermann** [comment]
Looks good. @Valodim : merge?

**Comment by Valodim** [comment]
Yup. I tested this before and code looks good now.

**Pull request merged**